### PR TITLE
Jython 3 MVP and Roadmap

### PR DESCRIPTION
**Pull request by jeff5**
Some ideas on MVP collated from our conversations, and an attempt at a roadmap.

### CODE REVIEW

**Comment by jeff5** [comment]
Ruby (or Jekyll) seem unhappy with my available machine(s), so I haven't been able to test these changes, but the menu changes seem low-risk and the Markdown renders ok on GitHub.

Anyway, It's the content I'd like to focus on, here or on the list.

File: jython-3-mvp.md
Comment (Stewori) [inline jython-3-mvp.md:175]: I once explored this approach, see https://github.com/Stewori/LiJy-launch (and consider to link it here).
However, it incorporates some (non-trivial) code taken from the Java-launcher source, resulting in GPL license with classpath exception.

File: jython-3-roadmap.md
Comment (Stewori) [inline jython-3-roadmap.md:212]: Maybe mention ctypes, cffi and a cython-backend here.

File: jython-3-mvp.md
Comment (jamesmudd) [inline jython-3-mvp.md:73]: Mac desktop

File: jython-3-mvp.md
Comment (jamesmudd) [inline jython-3-mvp.md:235]: Haven't looked into how possible it is but would be nice to avoid use of `jffi`/`jnr` to run completely on top of Java if possible this will make cross platform/architecture support much easier.

File: jython-3-roadmap.md
Comment (jamesmudd) [inline jython-3-roadmap.md:76]: Like the idea to add benchmarks, probably JMH, from as early as possible, it's these small core operations that will ultimately dominate real world performance, even if not much effort is directed to improve them at first, its good to have the framework there so when there is effort to focus on performance it can be well directed and measurable.

File: jython-3-roadmap.md
Comment (jamesmudd) [inline jython-3-roadmap.md:244]: Not sure if we should  be trying to cleanup the existing history somehow. The main thing that I would like changed if the history of checked in external libs, This would not be needed with a Gradle build, but keeping it in the history bloats the repo. But its always a trade off when considering rewriting history... https://git-scm.com/docs/git-filter-branch can do this and removing the `extlibs` dir should be enough.

**Review by jamesmudd — CHANGES_REQUESTED** [review]
Thanks for this I think it really helpful to publish this. Some minor comments but the navbar bit needs to be fixed before merging. Maybe just have 2 links for now, I think sub menus wouldn't be too hard but its probably not worth worrying about now. If you enable GH pages on your fork they will render a page you can view to check the content.

File: jython-3-mvp.md
Comment (jeff5) [reply jython-3-mvp.md:175]: Gosh that's complicated, and it worries me because you know a lot more than I about the native interface.

In my ignorance, I would have followed this: https://www.developer.com/java/data/how-to-create-a-jvm-instance-in-jni.html, where the target class would be org/python/util/jython, letting Jython do all the argument processing. The messiness is recompilation for each platform.

However, there is jlink to fall back on and I'm suggesting a native launcher is not MVP.

File: jython-3-roadmap.md
Comment (jeff5) [reply jython-3-roadmap.md:212]: I can mention them, but what would we like actually to say?

More broadly, I know calling modules written in  C would be useful, but until ideas of the pure Java core settle, I find it difficult to say what's involved. So not MVP, but interested in your thoughts now and when that settling has happened.

File: _data/navbar.yml
Comment (jeff5) [reply _data/navbar.yml:41]: Ah, ok. Thanks. Not as low risk as I thought. Very glad you were able to look in on it.

I'll flatten as suggested.

File: jython-3-mvp.md
Comment (jeff5) [reply jython-3-mvp.md:9]: You're right, I was being lazy.

File: jython-3-mvp.md
Comment (jeff5) [reply jython-3-mvp.md:50]: I agree in general for MVP. Long-term, I'd like scientific computing to be a proposition without C extensions, if we can manage it.

By some measures, Jython 2 is somewhat faster than CPython, although start-up definitely gives the opposite impression. Full use of the JVM's dynamic language features in JRuby had a massive effect on performance. In my toy, and almost certainly in Jython 3, the byte code interpreter uses `MethodHandle` but not `CallSite`. Generated JVM code would use ` invokedynamic` call sites fully, and I'm hoping for something good.

There's a real danger in setting any kind of target that we spend too much time micro-optimising the byte-code interpreter, when a JVM language is the goal. At the same time, if the basic operations of the abstract object API are much slower than in CPython, we're probably laying the wrong foundation. I could be talked down from the particular numbers.

In some simple loops and calls in my toy version I observe about 40% the speed of CPython. It's a bit disappointing, considering how succinct the Java is, but I'm resisting the temptation to delve further.

File: jython-3-mvp.md
Comment (jeff5) [reply jython-3-mvp.md:73]: Yes.

File: jython-3-roadmap.md
Comment (jeff5) [reply jython-3-roadmap.md:76]: I used JMH to check my work on some ` PyString` changes, and when I look at the old project, can't remember what it all means! But yes, JMH is what I meant. Useful as a canary in the coalmine.

File: jython-3-roadmap.md
Comment (jeff5) [reply jython-3-roadmap.md:244]: It hardly recommends itself on the man-page 😁 I've already lost several weekends to high-stakes git history wrangling.

You can address this with options to clone. This got me just 16MB of download:

    git clone --depth 4 --branch 3.8-scorched-earth https://github.com/jeff5/jython.git jython-scorched

I don't know if there's a way to make something like that the default experience.

Comment (jeff5) [reply jython-3-roadmap.md:244]: Afterthought: I guess the concern is for the CI? I think the same thing is possible, and for Travis a depth of 50 is used by default, but worth checking.

File: jython-3-mvp.md
Comment (jeff5) [reply jython-3-mvp.md:237]: It's what I meant here:
https://github.com/jython/jython.github.io/pull/21/files#diff-371cb2747f933eaf21de8ddd866028e5R216
Although I may be being too pure.

File: jython-3-mvp.md
Comment (Stewori) [reply jython-3-mvp.md:177]: The procedure described in https://www.developer.com/java/data/how-to-create-a-jvm-instance-in-jni.html is also at the core of java.exe and LiJyLaunch. However, it appears that java.exe has to do a lot of sanetizing before args are handled to the JNI api. E.g. jdb mode requires args in a special form, wildcard processing is done in the launcher, some system properties are detected, some memory stuff is configured and a bunch of things I do not understand. LiJy inherits this annoying complexity from java.exe's source code. Since the jython.py launcher used to call java.exe, taking this stuff from java.exe source was the only feasile way to provide the same behaviour regarding support for JVM command line args as in jython.py. That java.exe was from Java 8, maybe things changed since then, I never had a look again.

But actually, taking the source from java.exe is a rather painless solution to achieve this goal except for license issues. The easiest way would be to write a dual license, i.e. having the launcher licensed separately. GNU with classpath exception is not too bad. Or we can ask Oracle for permission to relicense this particular code. It is not very much code and not especially genuine. It rather and annoying useless work to rewrite/reinvent it just to escape the license.

However, it may be possible to link the required code dynamically from java.exe, see https://stackoverflow.com/questions/39131185/can-a-file-both-be-an-executable-exe-and-a-dynamic-link-library-dll-at-the-s. This would also be preferable for maintenance against upstream changes of java.exe. (I don't know if this approach is actually applicable to java.exe.)
Then jython.exe would link it dynamically and might not be subject to the viral license. I don't know if classpath exception also applies to dynamic linking, but otherwise dynamic linking of JNI would be subject to GNU with cp exception as well. So the approach in https://www.developer.com/java/data/how-to-create-a-jvm-instance-in-jni.html would inherently require relicensing then.

File: jython-3-roadmap.md
Comment (Stewori) [reply jython-3-roadmap.md:212]: Calling C by pure Java code won't be feasible until Project Panama lands. Currently one inherently has to write platform dependent C glue code or use something like JNA or JNR which is just the same with the only difference is that it was written by others. ctypes, cffi and cython are JNA/JNR-like libs for Python. Having Jython-versions would enable tons of Python modules at a time. This is still not sufficient for NumPy and SciPy but is a requirement. All three are used by NumPy and SciPy. So if you want to support scientific code, getting ctypes, cffi and cython working would be the first step.
That said, each of these is a huge project on its own. So it is likely a better approach to support [https://github.com/hpyproject/hpy](HPy) and hope that these libraries will be made HPy compatible as well. I know that cython announced HPy support. But then, supporting HPy is not feasible with pure Java code either.

File: jython-3-mvp.md
Comment (Stewori) [reply jython-3-mvp.md:51]: > Long-term, I'd like scientific computing to be a proposition without C extensions, if we can manage it.

If you talk about Java-ports of NumPy and SciPy, I estimate each one is a project comparable in complexity to Jython itself. In addition one still has to link blas and lapack, which inherently require C glue code. This can be based on existing projects like jblas, but then again, it still uses platform-dependent C code, just written by others. Project Panama may solve this but AFAIK it has not even a planned release date.

**Review by jamesmudd — APPROVED** [review]
Think it looks good to me, but maybe worth waiting on another review. Thanks for writing this up.

**Comment by jeff5** [comment]
Long enough for comments, I think.

**Pull request merged**
